### ISSUE TITLE
Reconsidering branch strategy

### ISSUE THREAD

**Issue by shu-mutou** [kubernetes/website#25806]
**This is a Feature Request**

**What would you like to be added**

Reconsidering branch strategy regarding with k8s release cycle and localization process.
https://kubernetes.io/docs/contribute/localization/#source-files
https://kubernetes.io/docs/contribute/localization/#branching-strategy

**Why is this needed**

* https://github.com/kubernetes/website/pull/25795#discussion_r548518703

+1 for creating `release-1.xx` for the current release.
The `master` branch used for for next release usually, but not for released version.

**Comments**

See also followings:
https://github.com/kubernetes/website/pull/25795
https://hackmd.io/smrEVJCJRee3cvXTIQ7bJA

**Comment by sftim** [comment]
> The `master` branch is not so suitable for documentation or translation for latest release that have already released, I think.

Hi @shu-mutou 

I didn't understand this sentence - can you rephrase it for clarity?

**Comment by sftim** [comment]
BTW, branch strategy for localizations is _per localization_; the approach that a “downstream” localization team takes is up to them, so long as the changes end up deployed.

**Comment by shu-mutou** [comment]
I agree with creating a `release-x.xx` branch for current release. L10n teams would be getting easier to create the milestone branch and to merge it to source branch.

**Comment by sftim** [comment]
> I agree with creating a release-x.xx branch for current release. L10n teams would be getting easier to create the milestone branch and to merge it to source branch.

The effect of merging _release-x.xx_ for the current branch would, the way I picture it, be identical to merging in the primary branch (_master_, or potentially _main_). What differences did you envisage @shu-mutou?

> We should clean up stale branches like dev-x.xx.

+:1: Once we've released a version, we don't need to keep the _dev-x.xx_ branch we used to prepare for that release.

**Comment by seokho-son** [comment]
> I agree with creating a `release-x.xx` branch for current release. L10n teams would be getting easier to create the milestone branch and to merge it to source branch.
> Also, we should clean up stale branches like `dev-x.xx`.

Hello @shu-mutou :) 
I am not sure creating `release-x.xx` branch for the current k8s release is beneficial to l10n teams. 
L10n teams can use `master` branch (the latest documents of the current k8s release.) as a base branch of localization milestone branch unless some of teams are targeting localizing documents that is not actively maintained.

In my understanding, @sftim 's idea to generate `release-x.xx` for the current branch as a mirror of `master` branch is to represent what is the branch for current release explicitly. (but it needs to be automated as @sftim mentioned, setting up a GitHub Action)

**Comment by shu-mutou** [comment]
Hi @sftim and @seokho-son 😄 
In my understanding, the benefit for creating `release-x.xx` branch is not only clarifying the branch for current release. If we have `release-x.xx` for the current release, l10n teams don't need to rebase their development branch from `master` even if they could not make it in time for merging it to `master` before the release, IMHO. 

What I'm most wondering about is that I'm using the `master` branch for the released version.
It's common to use the `master` branch for the next version, so I'm confused.
We have some tasks for `Docs Lead` in release cycle. https://github.com/kubernetes/sig-release/tree/master/releases/release-1.20#timeline
So I think we can create `release-x.xx` branch in same time of k8s release.
Also, I drew the branch strategy that I imagined as below.
https://hackmd.io/smrEVJCJRee3cvXTIQ7bJA#The-branch-strategy-I-had-imagined

**Comment by seokho-son** [comment]
```
What I'm most wondering about is that I'm using the master branch for the released version.
It's common to use the master branch for the next version, so I'm confused.
```
@shu-mutou It seems main issue here is changing usage of master branch :) 
I guess we don't need to include discussions for changing l10n development branch here. It seems a secondary issue.

I think sig-docs have a good reason for using master branch as documentation of current release. (because this repo is about documentation, and master branch is a branch what people may check first ? ) I am not a right person to discuss this issue :) Maybe we can invite other experts such as sig-docs leads to here or discuss it in weekly meetings.

Have nice holidays @shu-mutou :)

**Comment by oomichi** [comment]
> +1 for creating release-1.xx for the current release.
> The master branch used for for next release usually, but not for released version.

+1 for the above idea.
That is exactly common usage of branches of developments.

In addition, can I know why we need to have separated branches for each localization team as https://kubernetes.io/docs/contribute/localization/#branching-strategy ?
The files are separated with each directory under `./content` like e  en  es  fr  hi  id  it  ja and so on.
So the translation works cannot be interrupted between localization teams.

Actually I faced an issue related to current branch strategy as https://github.com/kubernetes/website/pull/24451#issuecomment-706491522
Japanese localization team said they will create their own branch for v1.20 and I need to wait for it. However that doesn't happen because they are so busy. If having a single branch for whole localization for each released version, we can reduce this kind of effort.
Then I'd like to know what is the point we have separated branches for each localization today.

**Comment by tengqm** [comment]
>  If having a single branch for whole localization for each released version, we can reduce this kind of effort.

That is too ideal a case for an imperfect world. Not all localization teams are equally staffed or actively engaged in localization. Each localization team can make their own decisions on how to track the upstream changes.

> I think sig-docs have a good reason for using master branch as documentation of current release. (because this repo is about documentation, and master branch is a branch what people may check first ? ) 

Yes. Changes to master are published right away. Take the branches today as an example:

- The master branch acts as the "latest version of the latest release". It is the default version you will see on k8s.io/docs.
- Changes to master will be archived in batches into the "release-1.20" branch, which is an "archived version of the latest release".
- Changes to master are picked up into "dev-1.21" as well, so no changes will get lost when 1.21 is out.
- Some changes may get merged into "release-1.19", "release-1.18", etc. which are "archived version of older releases".
- When kubernetes 1.21 is out, dev-1.21 becomes the new "master" and a "release-1.21" is created for archives purpose.

**Comment by oomichi** [comment]
@tengqm Thanks for your response.

> Each localization team can make their own decisions on how to track the upstream changes.

I agree with the above comment. It is nice to have their own decisions on how to track changes.
However I don't get the point why we need to have separated branches for each localization for doing that.
I guess we can know upstream changes by `git log ./content/en/` and know current progress of their own changes by `git log -p ./content/ja` for example even if having the same released branch for whole localization.
I'd like to know the reason I miss here.

**Comment by sftim** [comment]
> why we need to have separated branches for each localization for doing that.

In fact, only some (ie, not all) localizations choose to work in separate branches.

**Comment by shu-mutou** [comment]
@tengqm @sftim 
I'd like to add more descriptions to my PR (#25795), so could you let me know followings?
* When is the docs for current release archived on the release schedule? https://github.com/kubernetes/sig-release/tree/master/releases/release-1.21
* Can we translate `dev-1.21` branch for next version? e.g. can we create and merge `dev-1.21-ja.1` branch to it?

**Comment by tengqm** [comment]
@shu-mutou The release branch (e.g. `release-1.21`) will be created on the release week, which is usually week 12 of a release cycle.

I don't have an opinion whether creating `dev-1.21-ja.x` is a good idea. It is completely up to the localization team I think. If you decide to create such a branch, you may want to ensure changes in that branch will be merged back to the dev-1.21 timely.

**Comment by shu-mutou** [comment]
@tengqm Thank you for your reply. I'll add descriptions for these. 😉
